File: ios/Classes/Codec.swift
//
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0
//

import Foundation

struct Codec {
  private static let jsonEncoder = JSONEncoder()
  private static let jsonDecoder = JSONDecoder()
  
  static func encode<T>(result: Result<T>) -> String {
    return String(data: try! jsonEncoder.encode(result), encoding: .utf8)!
  }
  
  static func decodeInt(from arguments: Any?) -> Int {
    return arguments as! Int
  }
  
  static func decodePermission(from arguments: Any?) -> Permission {
    return Permission(rawValue: arguments! as! String)!
  }
  
  static func decodeLocationUpdatesRequest(from arguments: Any?) -> LocationUpdatesRequest {
    return try! jsonDecoder.decode(LocationUpdatesRequest.self, from: (arguments as! String).data(using: .utf8)!)
  }
}

File: android/src/main/kotlin/app/loup/geolocation/Codec.kt
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0

package app.loup.geolocation

import app.loup.geolocation.data.*
import com.squareup.moshi.Moshi

object Codec {

    private val moshi: Moshi = Moshi.Builder()
            .add(Permission.Adapter())
            .add(Priority.Adapter())
            .add(LocationUpdatesRequest.Strategy.Adapter())
            .build()

    fun encodeResult(result: Result): String =
            moshi.adapter(Result::class.java).toJson(result)

    fun decodeInt(arguments: Any?): Int =
            arguments!! as Int

    fun decodePermission(arguments: Any?): Permission =
            Permission.Adapter().fromJson(arguments!! as String)

    fun decodeLocationUpdatesRequest(arguments: Any?): LocationUpdatesRequest =
            moshi.adapter(LocationUpdatesRequest::class.java).fromJson(arguments!! as String)!!

    fun decodePermissionRequest(arguments: Any?): PermissionRequest =
        moshi.adapter(PermissionRequest::class.java).fromJson(arguments!! as String)!!

}
